pkgs/by-name/xa/xar/patches/0020-Fall-back-to-readlink-on-Linux.patch

```diff
@@ -0,0 +1,43 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Randy Eckenrode <randy@largeandhighquality.com>
+Date: Wed, 30 Oct 2024 20:19:20 -0400
+Subject: [PATCH 20/20] Fall back to readlink on Linux
+
+---
+ xar/lib/archive.c | 16 ++++++++++++++++
+ 1 file changed, 16 insertions(+)
+
+diff --git a/xar/lib/archive.c b/xar/lib/archive.c
+index b7f9cbf..a698f40 100644
+--- a/xar/lib/archive.c
++++ b/xar/lib/archive.c
+@@ -507,10 +507,26 @@ xar_t xar_fdopen_digest_verify(int fd, int32_t flags, void *expected_toc_digest,
+ 	// If there are hardlinks, the path we pick is the most recently opened by
+ 	// the filesystem; which is effectively random.
+ 	char path_buff[PATH_MAX];
++#if defined(__APPLE__)
+ 	if (fcntl(fd, F_GETPATH, path_buff) < 0) {
+ 		close(fd);
+ 		return NULL;
+ 	}
++#elif defined(__linux__)
++	// Linux has to get the path to the file via `/proc`.
++	if (readlink("/proc/self/fd/<n>", &path_buff, PATH_MAX) < 0) {
```

Comment:
This `<n>` cannot possibly be right, right?

Reply:
No, it’s not. It was a placeholder I forgot to convert because I didn’t have much time tonight to work on stuff. I’ve marked this draft until I can get a chance to fix it tomorrow.

Reply:
I updated it to use `snprintf` to generate the path to pass to `readlink`.